Some minor tweaks to the Commands API

zml2008 opened the pull request:
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/3262)

This is a companion to the Sponge PR where I spent some time looking through the commands. My main goal here was to add a bit to some documentation that felt unclear and look at some other minor things as I came across them.

Some more detail is present on the Sponge side of this PR.

zml2008 committed:
commands: JSON is actually SNBT

zml2008 committed:
command: Add to the documentation a bit

zml2008 committed:
command: Add utility methods for CommandTreeNode

zml2008 committed:
command: JSON -> SNBT, ArgumentReader edition

zml2008 commented on src/main/java/org/spongepowered/api/command/parameter/managed/clientcompletion/ClientCompletionTypes.java:
```diff
@@ -51,9 +51,13 @@
+     * keys, and can unambiguously distinguish every NBT tag type.</p>
      */
-    public static final DefaultedRegistryReference<ClientCompletionType> JSON = ClientCompletionTypes.key(ResourceKey.sponge("json"));
+    public static final DefaultedRegistryReference<ClientCompletionType> SNBT = ClientCompletionTypes.key(ResourceKey.sponge("snbt"));
```
Unsure about how to refer to this. While SNBT is a name for the format, and used on the MC wiki, the closest API type would be `DataContainer` -- would it make sense to call it `DATA_CONTAINER` instead?

gabizou replied:
I wouldn’t say datacintainer is quite up to snuff, it does not currently restrict that offering anything is within the limits of being compatible with a tag compound (some parts are translated specially). But it’s been on my todo list to revamp the api to be closer to that of having DataView like a ConfigNode while it’s implementation being closer to tags.

dualspiral replied:
Ideally, it shouldn't refer to NBT. I'd probably just call it "DATA" because that's what it is, it _is_ a data blob.

zml2008 replied:
I think portraying the format as anything as vague as `DATA` provides a false sense of abstraction. The client does parse this as a SNBT compound for the purposes of completion, so it cannot be any arbitrary string, and knowning the format used is important for plugins understanding the reasons for using it. I think this could *become* something like `DATA_CONTAINER` if gabe has the time to make the changes he wants to though. For now I'll keep it as `SNBT` since that is, while not ideal, not wrong, but I'm not opposed to changing it in the future.

zml2008 committed:
command: CommandManager is not globally accessible

Since command state is now tracked as part of datapack loading,
we have to refresh the commands map with every reload. This removes
the concept of a global commands manager from the API.

zml2008 committed:
command: JD fixes

zml2008 committed:
command: Expose an argument reader to Raw commands as well

dualspiral commented on src/main/java/org/spongepowered/api/event/lifecycle/RegisterCommandEvent.java:
```diff
@@ -63,6 +66,19 @@
+     *
+     * @return the command registrar
+     */
+    CommandRegistrar<C> registrar();
```
Why can't we get the registrar out of the server's `CommandManager` instead? As registrars are scoped to the datapack lifecycle, encouraging a plugin to hold on to a registrar is a bad idea.

I'm not 100% against exposing the registrar in the event, but this can't be the reason to do so.

`CommandRegistrar`s, as a result, should be `@DoNotStore`d.

dualspiral commented on src/main/java/org/spongepowered/api/command/Command.java:
```diff
@@ -95,7 +96,7 @@ static Builder builder() {
      * @throws CommandException Thrown on a command error
      */
-    CommandResult process(CommandCause cause, String arguments) throws CommandException;
+    CommandResult process(CommandCause cause, ArgumentReader.Mutable arguments) throws CommandException;
```
A comment rather than a request:

I'm very "eh" on this. I'm not really in favour of forcing the use of the reader on commands, but on the other hand, I understand the reasons why you'd want to provide it. I'd prefer being to create a reader from a string for use elsewhere too (if someone wanted to parse a chat message or something, for example), and keep this as a string so developers can decide, but it's a design/preference sort of thing here. 

I don't care enough to argue about it though because this may very well be the preferred direction anyway. I just see "Raw" as being as low level as possible, which this effectively prevents, but as you can get the string out of the reader, it's not _really_ an issue.

zml2008 replied:
hm, this was something I was unsure about as well at first. Why I eventually settled on exposing the argument reader as a method parameter was that we already allow Raw commands to generate a command tree, and an argument reader is an easy way to parse input in a way that is compatible with the types usable for command trees.

Because the raw string is available in `arguments.getRemaining()` this doesn't lock raw commands out of any ways of working, but does give them a pointer towards the right direction to go when seeking low-level access.

zml2008 committed:
Move command registrar access to be via command manager
